### PR TITLE
rfc: introduce dropout primitive

### PR DESCRIPTION
**Pull request by piotrchmiel**
[Rendered document](https://github.com/piotrchmiel/oneDNN/blob/dropout/rfcs/20200619-dropout/README.md)

RFC concerns introduction of dropout primitive in oneDNN. Full motivation is described inside document.

### CODE REVIEW

File: rfcs/20200619-dropout/README.md [hunk 1]
```diff
@@ -0,0 +1,425 @@
+# Proposal for DNNL dropout primitive
```
Comment (TaoLv) [inline rfcs/20200619-dropout/README.md:1]: You may want to use oneDNN instead of DNNL.

File: rfcs/20200619-dropout/README.md [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+## Introduction
+
+The main motivation is request from Amazon mxnet community:
```
Comment (TaoLv) [inline rfcs/20200619-dropout/README.md:5]: nit: Apache MXNet community

File: rfcs/20200619-dropout/README.md [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+usually default value is 0.5,
+
+where scale is 1 / (1 - rate)
+
+#### MXNET
+
+```python
+mxnet.gluon.nn.Dropout(rate, axes=(), **kwargs)
+```
+**Parameters**
```
Comment (TaoLv) [inline rfcs/20200619-dropout/README.md:41]: Here is a more detailed description of MXNet Dropout operator: https://mxnet.incubator.apache.org/api/python/docs/api/symbol/symbol.html#mxnet.symbol.Dropout

File: rfcs/20200619-dropout/README.md [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+parameter and if in the future there will be request to support that, it will
+require either add some kind v2 version in api or abi break.
+
+2. Should we support inplace computation (pytorch) ?
+
+In our opionion we can support that, it is not big implementation effort.
+
+3. Do we understand how differences in RNGs affect training results ?
+
+4. Where in DNN is RNG except dropout ?
```
Comment (TaoLv) [inline rfcs/20200619-dropout/README.md:422]: https://mxnet.cdn.apache.org/api/python/docs/api/ndarray/random/index.html#module-mxnet.ndarray.random

File: rfcs/20200619-dropout/README.md (hunk 2 above)
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md]: Fixed.

File: rfcs/20200619-dropout/README.md (hunk 1 above)
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md]: Fixed.

File: rfcs/20200619-dropout/README.md (hunk 4 above)
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md:422]: Thank you for answer :-) Maybe I missed something from you reply or question is not well formed. I mainly meant in what real deep neural network topologies random generators are used, in particular in what popular primitives. It is quite obvious that RNG it is used in various random like functions and distribution generators. I wanted to know practical usage of different frameworks random modules in applications other than dropout. I received already answer from frameworks team and I will update RFC in further commit.

File: rfcs/20200619-dropout/README.md (hunk 3 above)
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md:41]: In further commit I will attach note with link,

File: rfcs/20200619-dropout/README.md [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+      /// descriptor. Must be #dnnl_dropout.
+      dnnl_primitive_kind_t primitive_kind;
+      // The kind of propagation. Possible values: #dnnl_forward_training, #dnnl_backward
+      dnnl_prop_kind_t prop_kind;
+      // Source and destination memory descriptor.
+      dnnl_memory_desc_t data_desc;
+      // Source and destination gradient memory descriptor.
+      dnnl_memory_desc_t diff_data_desc;
+      // Seed memory descriptor.
+      dnnl_memory_desc_t seed_desc;
```
[the remark this answers is not in the document]
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md:136]: It is memory descriptor, seed will be passed at execution time. The same way it is done with tensors -> descriptors are passed in primitive constructors, memory with data in execution.

File: rfcs/20200619-dropout/README.md [hunk 6]
```diff
@@ -0,0 +1,425 @@
...
+Let's consider https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/dropout-inl.h
+since line 90. In each time they are creating new stream vsl stream and use
+viRngBernoulli. They don't save stream and viRngBernoulli doesn't save any state.
+The only thing that they're doing is gnenerating new seed in each param.
+
+Possible solution:
+- User should pass seed in execute instead primitive_descriptor (in intial draft,
+seed was passed in the primitive descriptor)
+- User passes seed in execute using DNNL_ARG_SEED. Primitivie in
+DNNL_ARG_SEED memory returns end-state of rng engine. In next calls of 'execute'
```
[the remark this answers is not in the document]
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md:405]: Based on statement above user is allowed to pass new seed. There are two use cases:
- user wants to run dropout with completely new seed - by specifying DNNL_ARG_SEED user can pass to execute in each run seed whatever he wants;
- user in next execution wants to use rng starting from end-state of previous execution - we allowing to do that by returning end seed.
At the end user all time is deciding what seed is passed to execution part.

File: rfcs/20200619-dropout/README.md [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+
+With that said, we can introduce a following vectorized formula for computing
+vectors of random values (with utilization of SIMD instructions):
+
+```
+N - lenght of AVX vector
+
+// first iteration (i=0)
+seed_t[0:N] = [seed, seed, ..., seed]
+a_t[0:N] = [1, A % modulus, A^2 % modulus, ..., A^N % modulus]
```
[the remark this answers is not in the document]
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md:374]: Fixed.

File: rfcs/20200619-dropout/README.md [hunk 8]
```diff
@@ -0,0 +1,425 @@
...
+
+        generate((tid*5), (tid*5) + 5, mask, seed, p);
+    }
+
+    std::copy(mask.begin(), mask.end(), std::experimental::make_ostream_joiner(std::cout, ", "));
+}
+```
+
+Both single threaded and multithreaded versions will output exactly the same
+value. For example Bernoulli distribution implementation for gcc is defined
```
[the remark this answers is not in the document]
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md:316]: I totally agree with you.  
1. Production kernel will be implemented if possible as vectorized implementation in jit assembly described in chapter "Random number generation - vectorized implementation".
2. C++ implementation will be used mostly in benchdnn reference output calculations. Regarding different glibc implementations I'm aware of that and it is valid point. I just wanted to emphasize that writing distribution which is consistent with concept of standard library distribution is quite easy. To avoid issues with differen glibc writing internal implementation consistent with concepts from std will be needed.

**Comment by piotrchmiel** [comment]
@mgouicem Just to confirm few things:
1. Your idea is to fuse dropout to rnn primitives (https://oneapi-src.github.io/oneDNN/group__cpp__api__rnn.html) as post-op or attribute ? Correct ?
2. Currently in frameworks dropout is implemented as two elementwise (binary) operations. As I see https://oneapi-src.github.io/oneDNN/v2.0-beta03/dev_guide_attributes_post_ops.html post-ops are not defined for rnn primitives. You mentioned "then this limits the usage of some primitives, typically RNNs". I just wanted from you to confirm statement: "Currently RNNs are also limited because binary is not fused with them" ?

File: rfcs/20200619-dropout/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mgouicem) [reply rfcs/20200619-dropout/README.md:406]: I am not sure about the second use case since it will introduce a primitive state which can cause a few problems, especially with the use of the primitive cache. I would rather force the user to pass a seed, and default to some fixed value if the seed is not provided.

**Comment by mgouicem** [comment]
@piotrchmiel ,
sorry should have been clearer on the dropout attribute idea.
Some primitive  need to apply dropout between two computations internally (e.g. if we fuse layers/iteration in RNN, we might have to apply the dropout between layers/iterations inside the primitive execution).
If we offer only a dropout primitive, a user can no more fuse layers/iterations in the RNN primitive if they need to apply dropout.
So we will have to have an attribute eventually to support it for RNN (or any other primitive where applying dropout would prevent fusion). Note that it is not a post-op per se since it happens in the middle of the computation for RNN, not at the end.
Now, the suggestion is, why not introduce such an attribute right now, so that we have only one mechanism to support dropout instead of two (a standalone primitive + an attribute)?

File: rfcs/20200619-dropout/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (piotrchmiel) [reply rfcs/20200619-dropout/README.md:406]: 1. In our proposition user always must pass DNNL_ARG_SEED.
2. Primitive is not storing in it's memory any state between executions. Let's consider proposed MCG31m1. Starting seed (base on passed initial seed) for each thread is computed as described in document MCG31m1 vectorised algorithm. We want also to compute "end seed" and return it to user the same way as we're returning tensors. User is deciding what seed will pass for next execution. He can take returned value and pass as DNNL_ARG_SEED or generate new custom value. Why do you think that it will affect primitive cache ?

**Comment by piotrchmiel** [comment]
@mgouicem Ok, so if I understand correctly you're proposing attribute instead of post-ops, because you want to enable fused operation not only at the end of execution of the primitive, but also possibly in the middle where applicable. 
1. How it is implemented now ? For example if you need dropout in the middle of RNN primitive execution, how you deal with that ? You simply don't use RNN primitives ?
2. Why do you think standalone primitive is obsolete not complementary to attribute approach? Attribute will require injecting implementation of dropout in all primitives where user wants to apply it. As you mentioned sometimes somewhere in the middle, sometimes at the end. Attribute allows possible the best fuse method, but it requires implementation of injection inside all primitives where we will identify that user may wants to use dropout. What if the user wanted to use dropout in other than we identified ? Standalone primitive will allow him to do that.

**Comment by densamoilov** [comment]
@piotrchmiel, are we going to continue with this RFC or we can close it?

**Comment by piotrchmiel** [comment]
@densamoilov I think
